src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/2.diff.sql

```diff
@@ -6,24 +6,31 @@
     TagLevel can be 0, 1 or 2 to represent Instance, Series or Study level
```

Comment:
Regarding the description on why this change. Why do we do initialization in transaction and upgrade not?

Reply:
Good question! I think as per discussion with John and @YazanMSFT - we could run the schema initialization scripts in a transaction. Since schema initialization would happen when we are standing up a new account and the SQLServer database is empty which is not the case for migration.

But last friday, I found out that one of the dicom script(CREATE FULLTEXT CATALOG) can't run in a transaction as per t-sql so anyways we would need to remove logic of running the initialization scripts wrapped in a transaction from schema-manager tool and instead we would update the latest initialization script(2.sql) idempotent too. I will create another PR to address this soon.
Please let me know if you have any more thoughts/questions. I am happy to schedule a quick call. Thanks!

Reply:
Thanks for the details Richa, can we sync when Brandon is back next week on this. Putting If condition on every SQL object, how do we test it?

For the Fulltext catalog, we can move it into another SQL, which can be configured in the code to mention it is transactional or not.

Reply:
Without wrapping these in transactions, we need to be very careful about the changes that we make to the schema. I think this leads to more strict requirements where the server has to be able to run when the schema update has been partially applied.

One example of this is where you add/alter a new "NOT NULL" constraint on a column but the script fails before the stored procedures are updated. There are a few other scenarios where this is the case. 

I hesitate to make everything not part of a transaction if the only reason is full text. Are there other reasons to not use a transaction?

(this is an old documentation, could quickly find updated about what is not allowed in a transaction - https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms191544(v=sql.105)?redirectedfrom=MSDN).

Reply:
Thanks Brandon, I totally agree we would need to be careful while making changes to the schema. We have considered that the schema changes are additive and not destructive. And binaries are backward compatible at least for one lesser schema version.

Here are the two parts:
1. Schema Migration.
As John mentioned back in the day - the motivation behind removing the transaction was performance. 
Using his example:
BEGIN TRANSACTION
 
ALTER TABLE ...
 
CREATE INDEX ...
 
END TRANSACTION
 
The ALTER TABLE statement holds an exclusive lock on the table which will only be released at the end of the transaction. This if we have expansive index creation statements that follow, this can result in blocking for extended periods of time.
 
So we agreed to made scripts as idempotent instead of transactional. 
We could definitely add variant of having a transaction around the ALTER/CREATE TABLE statements, which should be relatively quick. Then we have one or more transactions around the slower CREATE INDEX or UPDATE statements

2. Schema Initialization:
The motivation behind removing the transaction from initialization is mainly non transactional script(FULL TEXT).
I am gearing towards a hybrid approach here i.e. update schema manager tool to not add transaction for initialization but add transaction in the x.sql scripts. So in dicom 2.sql(considering 2 is the latest schema version), just after FULL TEXT script, we could begin transaction and end at the last.

Reply:
As per yesterday's discussion, have put up a PR for initial guidelines for SQLServer migration scripts - https://github.com/microsoft/healthcare-shared-components/pull/240 and updated the 2.diff.sql migration script in this PR to add transaction based on the nature of the statements.